### PR TITLE
Feat/llm cost user policy

### PR DESCRIPTION
**Pull request by Saadha123**
## Purpose                                                                                                                                                         
  LLM API calls have widely varying costs depending on the provider, model, and                                                                                      
  usage pattern (prompt tokens, completion tokens, caching, reasoning, speed tier,                                                                                   
  etc.). The gateway had no way to compute or expose the monetary cost of an LLM                                                                                     
  call, which is a prerequisite for any downstream cost-based rate limiting,                                                                                         
  chargeback, or observability feature.                                                                                                                              

  ## Goals
  - Introduce a new `llm-cost` policy that calculates the USD cost of an LLM API
    call at response time and injects it as an `x-llm-cost` response header.
  - Support all major LLM providers with their specific pricing models: OpenAI,
    Anthropic, Gemini (Google AI Studio + Vertex AI), and Mistral.
  - Load pricing data from a bundled `model_prices.json` file so gateway admins
    can update pricing without code changes.

  ## Approach
  The policy runs exclusively in the response phase:

  1. **Model identification** — Parses `$.model` from the response body
     (OpenAI-compatible), falling back to `$.modelVersion` for Gemini native format.

  2. **Pricing lookup** — Looks up the model name in a pricing map loaded once at
     instantiation from a JSON file (`pricing_file` system parameter). Lookup is
     case-insensitive and handles model version suffixes.

  3. **Provider-specific normalisation** — A per-provider `Calculator` interface
     normalises the raw response body into a common `Usage` struct
     (prompt tokens, completion tokens, cache read/write tokens, reasoning tokens,
     etc.). Anthropic's calculator also reads the request body to detect the
     Anthropic speed parameter which affects pricing.

  4. **Cost calculation** — A generic calculator computes the base cost from the
     `Usage` struct and `ModelPricing` entry (supporting tiered rates, cache
     discounts, batch discounts, etc.). The provider calculator then applies any
     provider-specific adjustments (e.g. Gemini geo multipliers, Vertex AI router
     flat cost).

  5. **Header injection** — `x-llm-cost` is set to the USD cost formatted to 10
     decimal places. `x-llm-cost-status` is set to `calculated` or
     `not_calculated` to disambiguate a zero-cost result from a failed lookup.

  ## User stories
  - As a gateway operator, I want the monetary cost of every LLM API call to be
    calculated and available as a response header so that downstream policies (e.g.
    budget rate limiting) and observability tooling can consume it.
  - As a gateway administrator, I want to ship a pricing file with the gateway
    image and override it via config so that pricing updates do not require
    redeploying code.
  - As a platform engineer, I want cost calculation to be non-blocking — if a model
    is unknown or the response body is malformed, the request should pass through
    with `x-llm-cost: 0` and a warning logged, not be rejected.

  ## Release note
  New policy: `llm-cost`. Calculates the USD monetary cost of LLM API calls at
  response time and injects the result as an `x-llm-cost` response header (e.g.
  `0.0000423100`). Supports OpenAI, Anthropic, Gemini, and Mistral with
  provider-specific pricing logic including tiered rates, cache discounts, batch
  discounts, reasoning tokens, and geo/speed multipliers. Pricing data is loaded
  from a configurable JSON file bundled with the gateway image.

  ## Documentation
  N/A — documentation to be added in a follow-up PR.

  ## Training
  N/A

  ## Certification
  N/A — new feature, certification questions to be scoped once documentation is
  finalised.

  ## Marketing
  N/A

  ## Automation tests
  - Unit tests
    - `llm_cost_test.go` (2107 lines): covers all four provider calculators across
      a wide range of models, token counts, and pricing edge cases (tiered rates,
      cache hits, batch mode, reasoning tokens, speed tiers, zero-token responses,
      unknown models, malformed bodies).
    - Static test data in `testdata/model_prices.json` (6901 lines) covers all
      supported models across providers.
  - Integration tests
    - N/A — the policy has no external dependencies at runtime beyond the pricing
      file; unit tests provide full coverage of the response-phase logic.

  ## Security checks
  - Followed secure coding standards: yes
  - Ran FindSecurityBugs plugin and verified report: yes
  - Confirmed no secrets committed: yes

  ## Samples
  N/A

  ## Related PRs
  - `llm-cost-based-ratelimit` policy — consumes the `x-llm-cost` header produced
    by this policy to enforce monetary spending budgets.

  ## Migrations (if applicable)
  N/A — new policy with no existing state.

## Summary by CodeRabbit

* **New Features**
  * Adds an LLM cost policy that computes per-request LLM costs and injects x-llm-cost and x-llm-cost-status metadata. Supports OpenAI, Anthropic, Gemini, and Mistral with provider-specific normalization, tiered pricing, cache carve-outs, geo/speed multipliers, grounding/web-search fees, and modality pricing. Configurable pricing file.

* **Tests**
  * Large test suite covering pricing lookup, normalization, tiering, provider behaviors, end-to-end flows, and many edge cases.

### CODE REVIEW

File: policies/llm-cost/llm_cost.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+// GetPolicy instantiates a new LLMCostPolicy. It loads the pricing database
+// from the pricing_file system parameter once at route-instantiation time.
+func GetPolicy(
+	_ policy.PolicyMetadata,
+	params map[string]interface{},
+) (policy.Policy, error) {
+	pricingFile, _ := params["pricing_file"].(string)
+	if pricingFile == "" {
+		return nil, fmt.Errorf("llm-cost: pricing_file system parameter is required but not set")
```
Comment (Arshardh) [inline policies/llm-cost/llm_cost.go]: lets check the time taken.. also we can cache the file inputs

File: policies/llm-cost/llm_cost.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+		"prompt_tokens", usage.PromptTokens,
+		"completion_tokens", usage.CompletionTokens,
+		"cost_usd", finalCost,
+	)
+
+	return setCostHeader(finalCost, costStatusCalculated)
+}
+
+// setCostHeader returns a ResponseAction that sets x-llm-cost and x-llm-cost-status.
+func setCostHeader(costUSD float64, status string) policy.ResponseAction {
```
Comment (Arshardh) [inline policies/llm-cost/llm_cost.go]: lets add this as a shared context. refer to analytics

File: policies/llm-cost/pricing.go [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+	}
+	pm := make(map[string]ModelPricing, len(raw))
+	for key, msg := range raw {
+		var p ModelPricing
+		if err := json.Unmarshal(msg, &p); err != nil {
+			slog.Warn("llm-cost: skipping invalid pricing entry", "key", key, "error", err)
+			continue
+		}
+		pm[key] = p
+	}
```
[the remark this answers is not in the document]
Comment (Arshardh) [reply policies/llm-cost/pricing.go:76]: fix this when the file is loaded to verify keys

File: policies/llm-cost/pricing.go [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+		"vertex_ai-language-models",
+		"vertex_ai-chat-models",
+		"vertex_ai-code-chat-models",
+		"vertex_ai-vision-models",
+		"vertex_ai-embedding-models":
+		return &GeminiCalculator{}
+	case "mistral":
+		return &MistralCalculator{}
+	default: // "openai", "text-completion-openai"
+		return &OpenAICalculator{}
```
Comment (Arshardh) [inline policies/llm-cost/pricing.go:267]: lets check if we can add a common impl here

File: policies/llm-cost/llm_cost.go (hunk 1 above)
Comment (Saadha123) [reply policies/llm-cost/llm_cost.go]: Previously the pricing file was loaded once per route instantiation. Added a package-level cache keyed by file path, so now the file is read only once per unique path and all routes share the same in-memory map.

File: policies/llm-cost/llm_cost.go (hunk 2 above)
Comment (Saadha123) [reply policies/llm-cost/llm_cost.go]: Done

File: policies/llm-cost/pricing.go (hunk 4 above)
Comment (Saadha123) [reply policies/llm-cost/pricing.go:267]: Added to todo.

File: policies/llm-cost/pricing.go [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+		if pricing.CacheCreationInputTokenCostAbove200k > 0 {
+			r.cacheWrite5m = pricing.CacheCreationInputTokenCostAbove200k
+			// TODO: if Anthropic ever defines cache_creation_input_token_cost_above_1hr_above_200k_tokens,
+			// select it here. For now, the >200k write rate applies to both TTLs.
+			r.cacheWrite1h = pricing.CacheCreationInputTokenCostAbove200k
+		}
+	case tierTokens > 128_000 && pricing.InputCostPerTokenAbove128k > 0:
+		r.input = pricing.InputCostPerTokenAbove128k
+		r.output = pricing.OutputCostPerTokenAbove128k
+	}
```
[the remark this answers is not in the document]
Comment (Saadha123) [reply policies/llm-cost/pricing.go:390]: Done 
864c760

**Review by Arshardh — APPROVED** [review]
